File: app/src/main/java/com/byd/dashcast/fission/FissionTeardownPlan.java
package com.byd.dashcast.fission;

import java.util.Collection;

/** Executes the required per-package Layout teardown order while isolating package failures. */
public final class FissionTeardownPlan {

    public interface Operations {
        String moveToDisplay0(String packageName) throws Exception;
        boolean forceStopAndWait(String packageName) throws Exception;
        void releaseSlot(String packageName) throws Exception;
        void onStepError(String packageName, String step, Throwable error);
    }

    private FissionTeardownPlan() {}

    public static void run(Collection<String> packages, boolean keepVirtualDisplays,
                           Operations operations) {
        if (packages == null || operations == null) return;
        for (String packageName : packages) {
            if (packageName == null || packageName.isEmpty()) continue;
            try {
                operations.moveToDisplay0(packageName);
            } catch (Throwable error) {
                operations.onStepError(packageName, "move", error);
            }
            try {
                if (!operations.forceStopAndWait(packageName)) {
                    operations.onStepError(packageName, "force-stop-verify",
                            new IllegalStateException("process death not verified"));
                }
            } catch (Throwable error) {
                operations.onStepError(packageName, "force-stop", error);
            }
            if (!keepVirtualDisplays) {
                try {
                    operations.releaseSlot(packageName);
                } catch (Throwable error) {
                    operations.onStepError(packageName, "release", error);
                }
            }
        }
    }
}

File: app/src/main/java/com/byd/dashcast/fission/FissionTeardownPlan.kt
package com.byd.dashcast.fission

/** Executes the required per-package Layout teardown order while isolating package failures. */
object FissionTeardownPlan {

    interface Operations {
        @Throws(Exception::class)
        fun moveToDisplay0(packageName: String): String?
        @Throws(Exception::class)
        fun forceStopAndWait(packageName: String): Boolean
        @Throws(Exception::class)
        fun releaseSlot(packageName: String)
        fun onStepError(packageName: String, step: String, error: Throwable)
    }

    @JvmStatic
    fun run(packages: Collection<String?>?, keepVirtualDisplays: Boolean, operations: Operations?) {
        if (packages == null || operations == null) return
        for (packageName in packages) {
            if (packageName.isNullOrEmpty()) continue
            try {
                operations.moveToDisplay0(packageName)
            } catch (error: Throwable) {
                operations.onStepError(packageName, "move", error)
            }
            try {
                if (!operations.forceStopAndWait(packageName)) {
                    operations.onStepError(packageName, "force-stop-verify",
                        IllegalStateException("process death not verified"))
                }
            } catch (error: Throwable) {
                operations.onStepError(packageName, "force-stop", error)
            }
            if (!keepVirtualDisplays) {
                try {
                    operations.releaseSlot(packageName)
                } catch (error: Throwable) {
                    operations.onStepError(packageName, "release", error)
                }
            }
        }
    }
}
